packages/expo-asset/android/src/main/java/expo/modules/asset/AssetModule.kt

```diff
@@ -0,0 +1,89 @@
+package expo.modules.asset
+
+import android.net.Uri
+import expo.modules.interfaces.filesystem.Permission
+import expo.modules.kotlin.AppContext
+import expo.modules.kotlin.exception.CodedException
+import expo.modules.kotlin.functions.Coroutine
+import expo.modules.kotlin.modules.Module
+import expo.modules.kotlin.modules.ModuleDefinition
+import kotlinx.coroutines.withContext
+import java.io.File
+import java.io.FileInputStream
+import java.net.URI
+import java.security.DigestInputStream
+import java.security.MessageDigest
+
+internal class UnableToDownloadAssetException(url: String) :
+  CodedException("Unable to download asset from url: $url")
+
+class AssetModule : Module() {
+  private fun getMD5HashOfFilePath(uri: URI): String {
+    val md = MessageDigest.getInstance("MD5")
+    return md.digest(uri.toString().toByteArray()).joinToString("") { "%02x".format(it) }
+  }
+
+  private fun getMD5HashOfFileContent(file: File): String? {
+    return try {
+      val md = MessageDigest.getInstance("MD5")
+      val fis = FileInputStream(file)
+      DigestInputStream(fis, md).use { dis ->
+        while (dis.read() != -1) {
+          // Read the file content
+        }
+        md.digest().joinToString(separator = "") { "%02x".format(it) }
+      }
```

Comment:
not sure doing the read of the file content manually is required. Here's what we use in expo-updates:

```suggestion
      FileInputStream(file).use { inputStream ->
        DigestInputStream(
          inputStream,
          MessageDigest.getInstance("MD5")
        ).use { digestInputStream ->
          digestInputStream.messageDigest.digest().joinToString(separator = "") { "%02x".format(it) }
        }
      }
```